### ISSUE TITLE
Add warning category to WARNINGs on Data.List.{head,tail}

### ISSUE THREAD

**Issue by adamgundry** [haskell/core-libraries-committee#114]
Recently the CLC approved proposal #87 to add `WARNING` pragmas to `head` and `tail` from `Data.List`. Users who do not wish to see such warnings are advised to use `-Wno-warnings-deprecations`. However this has the effect of disabling *all* messages from `WARNING` pragmas, which is not necessarily what the user wants.

Partly motivated by this, and in order to provide more flexibility to users, the GHC Streering Committee has subsequently accepted https://github.com/ghc-proposals/ghc-proposals/pull/541 that permits `WARNING` pragmas to be annotated with a category name. In particular, it will permit the following:

```hs
{-# WARNING in "x-partial" head "This is a partial function...." #-}
{-# WARNING in "x-partial" tail "This is a partial function...." #-}
```

Users will then be able to disable these warnings with `-Wno-x-partial` (and similar flags, e.g. promoting them to an error with `-Werror=x-partial`). They are no longer under the control of `-Wwarnings-deprecations` (but there is a new warning group `-Wextended-warnings` that covers all messages from `WARNING` pragmas regardless of category).

I propose that the `x-partial` warning category be added to the `WARNING` pragmas being added to `head` and `tail` by [!9290](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9290).

This is no more of a breaking change than the original #87, provided it happens in the same GHC release. (If it happens in a subsequent release, users who set `-Wno-warnings-deprecations` will start to see warnings again until they set `-Wno-x-partial`.)

An implementation of warning categories in GHC is available at [!9684](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9684). It seems likely that this should be ready in time for GHC 9.8, which is also likely to be the first release that can include #87.

**Comment by Kleidukos** [comment]
Excellent suggestion!

### With my Haddock hat on: 

I'd be delighted to have such a mechanism that I could re-use from inside Haddock to signify the cause of warnings.
There is an adjacent thread in the Forum: https://discourse.haskell.org/t/collectings-occurences-of-incomplete-patterns-to-make-them-available-to-haddock/5494/


### With my user hat on:

I support the addition of a partiality tag for the WARNING pragma, and even go further:

* Add it to the Errors index;
* Have a strict, unambiguous meaning of "partiality" in this context that users can read on the errors index page.

**Comment by Bodigrim** [comment]
It would be nice to have a framework of softly recommended categories lest we end up with `x-partial`, `x-non-total` and `x-diverging`. But I don't know how to foster such catalog. Having an entry in the error index sounds like a good idea. 

Narrowing back to this ticket, can we agree on a CLC guidance for users which functions are suggested to use `x-partial`?

**Comment by jappeace** [comment]
It's a good idea,
but the claim this effects all warnings is an exaggeration since you can disable warnings per module.
For example adding `{-# OPTIONS_GHC -Wno-warnings-deprecations #-}` on the top of your module.

**Comment by adamgundry** [comment]
I've updated the issue to reflect a minor syntactic change in the design of the pragma. Please could the CLC make a decision on this?

I agree that it could be useful to define a catalogue of recommended categories, perhaps alongside the errors index, and to have guidance on when `x-partial` in particular is applicable. I don't believe that is necessary for this change, however.


> the claim this effects all warnings is an exaggeration since you can disable warnings per module.

The text refers to "disabling all messages from `WARNING` pragmas" which is precisely the effect of `-Wno-warnings-deprecations` (prior to the introduction of categories).

**Comment by hasufell** [comment]
Sounds good. Since I'm not a friend of warning about head/tail in the first place, anything that makes it less annoying is welcome.

**Comment by adamgundry** [comment]
@Bodigrim 

> Please could the CLC make a decision on this?

**Comment by parsonsmatt** [comment]
+1

**Comment by Bodigrim** [comment]
>  I don't believe that is necessary for this change, however.

OK, let's vote on the proposal as is. 

----

Dear CLC members, let's vote on the proposal to classify warnings for `Data.List.{head,tail}` as belonging to `partial` group, allowing a fine-grained control over them with `-Wno-x-partial`. This has been discussed originally in the course of #87, but the GHC proposal for custom warnings has not been approved yet. The implementation of the feature is in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9684, which I assume could be merged reasonably soon.

@tomjaguarpaw @chshersh @angerman @mixphix @hasufell 

----

+1 from me.

**Comment by mixphix** [comment]
+1

**Comment by konsumlamm** [comment]
> This has been discussed originally in the course of #87, but the GHC proposal for custom warnings has not been approved yet.

I assume you mean "the GHC proposal for custom warnings ha**d** not been approved yet"? It has been accepted since then.

**Comment by konsumlamm** [comment]
> More diagnostics sound good to me +1. It may break people with -Weverything -Werror, but then you are asking for it.

Note that `head` and `tail` already have these warnings (see #87), this proposal is about making it possible to control them with a new warning category. So this doesn't break anything that wasn't already broken before.

**Comment by hasufell** [comment]
+1

**Comment by tomjaguarpaw** [comment]
+1

**Comment by Bodigrim** [comment]
Approved unanimously, thanks all.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Authors**            |  @adamgundry  |
| **Status**             |  not implemented |
| **`base` version**     | `4.19.0.0`  |
| **Merge Request (MR)** | ??? |
| **Blocked by**         | ??? |
| **CHANGELOG entry**    | missing (please, [raise an MR to `base` to update changelog][changelog])  |
| **Migration guide**    |  not needed |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂  

<hr>

@adamgundry could you share a status update on the implementation and what are next steps?

I see that the relevant GHC feature was finally merged:

* https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9684

I believe, the next step would be to update the original MR with approved categories:

* https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9290
